### PR TITLE
1122 add or remove users in admin set

### PR DESCRIPTION
**Pull request by alishaevn**
Co-author: Alisha Evans <alisha@notch8.com>
Co-author: Dylan Salay <dylan@notch8.com>

# Ticket
https://app.zenhub.com/workspaces/yul-dc-5e50083561915b11fc9e5850/issues/yalelibrary/yul-dc/1122

# Summary
- users can be added as a viewer _or_ editor per admin set
- users can be removed from their role per admin set

# Demo
https://share.getcloudapp.com/4gu1E1PK

# Resources
- https://rubyinrails.com/2019/07/11/rails-routes-member-vs-collection/

### CODE REVIEW

**Review by mikeapp — COMMENTED** [review]
Sorry, looking at the wrong PR - will comment in 1121!

File: app/controllers/roles_controller.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  before_action :set_user
+  before_action :set_item, only: [:create]
+
+  def create
+    if @item && @user.has_role?(params[:role], @item)
+      redirect_back(fallback_location: root_path, flash: { alert: "User: #{@user.uid} is already assigned as #{params[:role]}" })
+    elsif @item
+      # each user only gets one role per item, remove all others first.
+      # This removes the role for all users who have that role, not only @user
+      # @user.roles.where(resource: @item).destroy_all
```
Comment (dylansalay) [inline app/controllers/roles_controller.rb]: This is destroying the role for all users who share that role, not just the user we are adding a new role for.

File: app/controllers/roles_controller.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      params[:role] == "viewer" ? @user.remove_role(:editor, @item) : @user.remove_role(:viewer, @item)
+      @user.add_role(params[:role], @item)
+      redirect_back(fallback_location: root_path, notice: show_notice(@user, params[:role]))
+    else
+      @user.add_role(params[:role])
+      redirect_back(fallback_location: root_path, notice: show_notice(@user, params[:role]))
+    end
+  end
+
+  # We tried using the standard destroy action but again, passing in the role id destroys the role for all users who are sharing that role, not just the user whose role we want to destroy
```
Comment (dylansalay) [inline app/controllers/roles_controller.rb]: The Rails standard destroy action of destroying a Role is the same problem we are seeing above. Users who are editors all share the same editor role. So destroying that role destroys the role for all users.

**Review by orangewolf — APPROVED** [review]

File: app/models/user.rb [hunk 3]
```diff
@@ -27,6 +28,10 @@ def sysadmin=(value)
     end
   end
 
+  def find_role(role, admin_set)
```
Comment (martinlovell) [inline app/models/user.rb:31]: This looks problematic and unused. You'd need resource_type to find a unique role.

**Pull request merged**